File: Android/shared/src/main/kotlin/shared/dto/enums/CustomerUser.kt
@file:UseSerializers(URISerializer::class)

package shared.dto.enums

import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import shared.dto.objects.User
import shared.dto.protocols.Responsable
import shared.serializers.URISerializer
import java.net.URI

@Serializable(CustomerUser.Serializer::class)
sealed class CustomerUser : Responsable {

    @Serializable
    data class Link(
        val link: URI,
    ) : CustomerUser()

    @Serializable
    data class Value(
        val value: User.Responses.Partial,
    ) : CustomerUser()

    internal object Serializer :
        JsonContentPolymorphicSerializer<CustomerUser>(CustomerUser::class) {
        override fun selectDeserializer(element: JsonElement): DeserializationStrategy<CustomerUser> {
            return when (element.jsonObject.contains("link")) {
                true -> Link.serializer()
                false -> Value.serializer()
            }
        }
    }
}

File: Node/src/dto/enums/CustomerUser.ts
import { User } from "../objects/User";
import { ValidateNested } from "class-validator";

export type CustomerUser = CustomerUser.Link   | 
                           CustomerUser.Value

export namespace CustomerUser {

    export class Link {
        link: URL

        constructor(link: URL) {
            this.link = link
        }
    }

    export class Value {
        @ValidateNested()
        value: User.Responses.Partial

        constructor(value: User.Responses.Partial) {
            this.value = value
        }
    }
}